tests/library/test_json_schema.py

```diff
@@ -0,0 +1,523 @@
+import json
+
+from typing import Any
+
+import pytest
+from jsonschema import validate
+
+from guidance import models
+from guidance.library import gen_json
+
+
+def to_compact_json(target: any) -> str:
+    # See 'Compact Encoding':
+    # https://docs.python.org/3/library/json.html
+    # Since this is ultimately about the generated
+    # output, we don't need to worry about pretty printing
+    # and whitespace
+    return json.dumps(target, separators=(",", ":"))
+
+
+def _generate_and_check(target_obj: Any, schema_obj):
+    # Sanity check what we're being asked
+    validate(instance=target_obj, schema=schema_obj)
+
+    prepared_string = f"<s>{to_compact_json(target_obj)}"
+    lm = models.Mock(prepared_string.encode())
+
+    # Run with the mock model
+    CAPTURE_KEY = "my_capture"
+    lm += gen_json(name=CAPTURE_KEY, json_schema=schema_obj)
+
+    # Make sure the round trip works
+    assert json.loads(lm[CAPTURE_KEY]) == target_obj
+
+
+def _check_failed_generation(bad_string: str, expected_output: Any, schema_obj):
+    prepared_string = "<s>" + bad_string
+    lm = models.Mock(prepared_string.encode())
+
+    # Run with the mock model
+    CAPTURE_KEY = "my_capture"
+    lm += gen_json(name=CAPTURE_KEY, json_schema=schema_obj)
+
+    assert json.loads(lm[CAPTURE_KEY]) == expected_output
+
+
+def test_null():
+    schema = """{"type": "null" }"""
+
+    target_obj = None
+
+    # First sanity check what we're setting up
+    schema_obj = json.loads(schema)
+    validate(instance=target_obj, schema=schema_obj)
+
+    _generate_and_check(target_obj, schema_obj)
+
+
+@pytest.mark.parametrize("target_obj", [True, False])
+def test_boolean(target_obj):
+    schema = """{"type": "boolean" }"""
+
+    # First sanity check what we're setting up
+    schema_obj = json.loads(schema)
+    validate(instance=target_obj, schema=schema_obj)
+
+    _generate_and_check(target_obj, schema_obj)
+
+
+class TestInteger:
+    schema = """{ "type": "integer" }"""
+
+    @pytest.mark.parametrize(
+        "my_int",
+        [0, 1, 100, 9876543210, 99, 737, 858, -1, -10, -20],
+    )
+    def test_integer_schema(self, my_int):
+        # First sanity check what we're setting up
+        schema_obj = json.loads(TestInteger.schema)
+        validate(instance=my_int, schema=schema_obj)
+
+        # The actual check
+        _generate_and_check(my_int, schema_obj)
+
+    @pytest.mark.parametrize(
+        ["bad_string", "expected_capture"],
+        [
+            ("9999a7777", 9999),
+            ("123, []", 123),
+            ("a321", 2),  # For some reason, '2' is the failure output from Mock
```

Comment:
These failure cases are irking me. Need to figure out what's happening with the `Mock`